backend/src/service/nextcloud.rs

```diff
@@ -0,0 +1,284 @@
+use crate::error::ServiceError;
+use actix_http::{Method, StatusCode};
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+const BASE_URL: &str = "https://cloud.permaplant.net/";
+
+#[derive(Debug, Deserialize)]
+pub struct CreateCircleResponse {
+    pub meta: Meta,
+    pub data: CircleData,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Meta {
+    pub status: String,
+    pub statuscode: i32,
+    pub message: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct CircleData {
+    pub id: String,
+    pub name: String,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct FindCircleData {
+    pub element: FindCircleDataElement,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FindCircleDataElement {
+    pub id: String,
+    pub user_id: String,
+    pub user_type: i32,
+    pub display_name: String,
+    pub instance: String,
+}
+
+fn extract_id_from_doc(doc: &str) -> String {
+    let res: CreateCircleResponse = quick_xml::de::from_str(doc).unwrap();
+    log::debug!("{:?}", res);
+    res.data.id
+}
+
+fn extract_id_from_circle_search_response(doc: &str) -> String {
+    let res: FindCircleResponse = quick_xml::de::from_str(doc).unwrap();
+    log::debug!("{:?}", res);
+    res.data.element.id.clone()
+}
+
+pub async fn setup_nextcloud_shares(
+    user_id: Uuid,
+    map_name: &str,
+    token: &str,
+    _should_share_publicly: bool,
+) -> Result<(), ServiceError> {
+    // create Nextcloud circle with the map name
+    let circle_id = create_nextcloud_circle(map_name, token).await?;
+    // create a directory with the map name
+    create_directory(user_id, map_name, token).await?;
+    // create a directory for the base layer within
+    create_directory(user_id, &format!("{}/Base/", map_name), token).await?;
+    share_directory(map_name, token, circle_id).await?;
+
+    // FIXME: nextcloud returns 500 error
```

Comment:
Here is still a little leftover because we don't know why nextcloud failed here.